lib/SPIRV/LLVMBitreverse.h

```diff
@@ -0,0 +1,134 @@
+//===- LLVMBitreverse.h - implementation of llvm.bitreverse -===//
+//
+//                     The LLVM/SPIRV Translator
+//
+// Copyright (c) 2024 The Khronos Group Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements lowering of llvm.bitreverse.* into basic LLVM
+// operations.
+//
+//===----------------------------------------------------------------------===//
+
+// The IR below is slightly manually modified IR which was produced by Clang
+// from the C code below.
+//
+// #include <stdlib.h>
+// #include <stdint.h>
+//
+// uint16_t llvm_bitreverse_i16(uint16_t a) {
+//   uint16_t ret=0;
+//   uint32_t i;
+//
+//   for (i=0; i<16; i++) {
+//     ret<<=1;
+//     if (a&0x1)
+//       ret|=1;
+//     a>>=1;
+//   }
+//   return ret;
+// }
+//
+// uint32_t llvm_bitreverse_i32(uint32_t a) {
+//   uint32_t ret=0;
+//   uint32_t i;
+//
+//   for (i=0; i<32; i++) {
+//     ret<<=1;
+//     if (a&0x1)
+//       ret|=1;
+//     a>>=1;
+//   }
+//   return ret;
+// }
+//
+// uint64_t llvm_bitreverse_i64(uint64_t a) {
+//   uint64_t ret=0;
+//   uint32_t i;
+//
+//   for (i=0; i<64; i++) {
+//     ret<<=1;
+//     if (a&0x1)
+//       ret|=1;
+//     a>>=1;
+//   }
+//   return ret;
+// }
+//
+// Clang options: -emit-llvm -O1 -g0 -fno-discard-value-names
+// -O1 is used instead of -O2 to avoid generating a call to llvm.bitreverse.i16
+
+static const char LLVMBitreverse[]{R"(
```

Comment:
That is an interesting approach which is worth exploring, but without tablegen it won't work.
1. There are many more integers, then i16/i32/i64;
2. The intrinsic function works not only with scalars, but with vectors as well.

Reply:
Added support for (i8/i16/i32/i64)x(scalar,vec-2,vec-4).